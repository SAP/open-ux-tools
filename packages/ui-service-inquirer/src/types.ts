import type { AbapDeployConfigAnswersInternal } from '@sap-ux/abap-deploy-config-inquirer';
import type { ServiceProvider, UiServiceGenerator } from '@sap-ux/axios-extension';
import type { Destination } from '@sap-ux/btp-utils';
import type { YUIQuestion } from '@sap-ux/inquirer-common';
import type { promptNames, SystemSelectionAnswerType } from '@sap-ux/odata-service-inquirer';
import type { BackendSystem } from '@sap-ux/store';
import type { Question } from 'inquirer';

export interface UiServiceAnswers extends AbapDeployConfigAnswersInternal {
    [promptNames.systemSelection]?: SystemSelectionAnswerType;
    businessObjectInterface?: string;
    abapCDSView?: string;
    objectType?: ObjectType;
    serviceName?: string;
    draftEnabled?: boolean;
    launchAppGen?: boolean;
}

export interface ServiceConfig {
    content: string;
    serviceName: string;
}

export type SystemSelectionQuestion = Question<UiServiceAnswers>;
export type ServiceConfigQuestion = YUIQuestion<UiServiceAnswers>;

export interface SystemSelectionAnswers {
    /**
     * The connected system will allow downstream consumers to access the connected system without creating new connections.
     *
     */
    connectedSystem?: {
        /**
         * Convienence property to pass the connected system
         */
        serviceProvider: ServiceProvider;

        /**
         * The persistable backend system representation of the connected service provider
         * `newOrUpdated` is set to true if the system was newly created or updated during the connection validation process and should be considered for storage.
         */
        backendSystem?: BackendSystem & { newOrUpdated?: boolean };

        /**
         * The destination information for the connected system
         */
        destination?: Destination;
    };
    objectGenerator?: UiServiceGenerator;
}

export enum ObjectType {
    BUSINESS_OBJECT = 'BusinessObject',
    CDS_VIEW = 'CDSView'
}

export interface ServiceConfigOptions {
    useDraftEnabled?: boolean;
    useLaunchGen?: boolean;
}
