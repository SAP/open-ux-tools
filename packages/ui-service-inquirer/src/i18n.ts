import type { TOptions } from 'i18next';
import i18next from 'i18next';
import translations from './translations/ui-service-inquirer.i18n.json';

const uiServiceI18nNamespace = 'ui-service-inquirer';
/**
 * Initialize i18next with the translations for this module.
 */
export async function initI18n(): Promise<void> {
    await i18next.init({ lng: 'en', fallbackLng: 'en' });
    i18next.addResourceBundle('en', uiServiceI18nNamespace, translations);
}

/**
 * Helper function facading the call to i18next. Unless a namespace option is provided the local namespace will be used.
 *
 * @param key i18n key
 * @param options additional options
 * @returns {string} localized string stored for the given key
 */
export function t(key: string, options?: TOptions): string {
    if (!options?.ns) {
        options = Object.assign(options ?? {}, { ns: uiServiceI18nNamespace });
    }
    return i18next.t(key, options);
}

initI18n().catch(() => {
    // Needed for lint
});
