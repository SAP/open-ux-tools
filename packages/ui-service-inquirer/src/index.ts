export * from './types';
export { getSystemSelectionPrompts, getConfigPrompts } from './prompts';
