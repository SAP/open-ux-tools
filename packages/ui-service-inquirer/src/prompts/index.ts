export * from './prompts';
export * from './prompt-helper';
