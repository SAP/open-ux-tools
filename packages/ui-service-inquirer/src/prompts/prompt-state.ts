import type { ServiceConfig, SystemSelectionAnswers } from '../types';

/**
 * Much of the values returned by the ui service inquirer prompting are derived from prompt answers and are not direct answer values.
 * Since inquirer does not provide a way to return values that are not direct answers from prompts, this class will maintain the derived values
 * across prompts statically for the lifespan of the prompting session.
 *
 */
export class PromptState {
    private static _systemSelection: SystemSelectionAnswers = {};
    private static _serviceConfig: ServiceConfig;

    /**
     * Returns the current state of the service config.
     *
     * @returns {ServiceConfig} service config
     */
    public static get systemSelection(): SystemSelectionAnswers {
        return this._systemSelection;
    }

    /**
     * Set the state of the system selection.
     *
     * @param {SystemSelectionAnswers} value - system selection value
     */
    public static set systemSelection(value: Partial<SystemSelectionAnswers>) {
        this._systemSelection = value;
    }

    /**
     * Returns the current state of the service config.
     *
     * @returns {ServiceConfig} service config
     */
    public static get serviceConfig(): ServiceConfig {
        return this._serviceConfig;
    }

    /**
     * Set the state of the service config.
     *
     * @param {ServiceConfig} value - service config value
     */
    public static set serviceConfig(value: ServiceConfig) {
        this._serviceConfig = value;
    }

    static reset(): void {
        PromptState.systemSelection = {};
        PromptState.serviceConfig = {
            content: '',
            serviceName: '',
            showDraftEnabled: false
        };
    }

    static resetConnectedSystem(): void {
        PromptState.systemSelection = {};
    }

    static resetServiceConfig(): void {
        PromptState.serviceConfig = {
            content: '',
            serviceName: '',
            showDraftEnabled: false
        };
    }
}
