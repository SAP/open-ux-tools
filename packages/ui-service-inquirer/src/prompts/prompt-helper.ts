import type { AbapServiceProvider } from '@sap-ux/axios-extension';
import { AbapCDSViewService, BusinessObjectsService } from '@sap-ux/axios-extension';
import { isAppStudio, type Destination } from '@sap-ux/btp-utils';
import { HELP_NODES } from '@sap-ux/guided-answers-helper';
import type { ValidationLink } from '@sap-ux/inquirer-common';
import { ERROR_TYPE, ErrorHandler, setTelemetryClient } from '@sap-ux/inquirer-common';
import type { BackendSystem } from '@sap-ux/store';
import type { AbapTarget, DestinationAbapTarget, UrlAbapTarget } from '@sap-ux/system-access';
import { ClientFactory } from '@sap-ux/telemetry';
import type { ListChoiceOptions } from 'inquirer';
import { t } from '../i18n';

/**
 * Get the business objects.
 *
 * @param provider - the provider to get the business objects
 * @returns the business objects
 */
export async function getBusinessObjects(provider: AbapServiceProvider): Promise<ListChoiceOptions[]> {
    const businessObjectsService = await provider.getAdtService<BusinessObjectsService>(BusinessObjectsService);
    const businessObjects = await businessObjectsService?.getBusinessObjects();
    return (
        businessObjects?.map((bo: any) => {
            return { name: `${bo.name} (${bo.description})`, value: bo };
        }) ?? []
    );
}

/**
 * Get the ABAP CDS views.
 *
 * @param provider - the provider to get the ABAP CDS views
 * @returns the ABAP CDS views
 */
export async function getAbapCDSViews(provider: AbapServiceProvider): Promise<ListChoiceOptions[]> {
    const abapCDSViewsService = await provider.getAdtService<AbapCDSViewService>(AbapCDSViewService);
    const abapCDSViews = await abapCDSViewsService?.getAbapCDSViews();
    return (
        abapCDSViews?.map((abapCDSView: any) => {
            return { name: `${abapCDSView.name} (${abapCDSView.description})`, value: abapCDSView };
        }) ?? []
    );
}

/**
 * Create an ABAP target from backend system or destination.
 *
 * @param destination - the destination
 * @param backendSystem - the backend system
 * @returns the ABAP target
 */
export function createAbapTarget(destination?: Destination, backendSystem?: BackendSystem): AbapTarget {
    if (isAppStudio()) {
        return { destination: destination?.Name } as DestinationAbapTarget;
    } else {
        return { url: backendSystem?.url, client: backendSystem?.client } as UrlAbapTarget;
    }
}

/**
 * Get the service name choices.
 *
 * @param serviceName - the service name
 * @returns the service name choices
 */
export function getServiceNameChoices(serviceName: string): ListChoiceOptions[] {
    return [
        {
            name: serviceName,
            value: serviceName
        }
    ];
}

/**
 * Get the validation error link.
 *
 * @returns the validation error link
 */
export async function getValidationErrorLink(): Promise<ValidationLink> {
    setTelemetryClient(ClientFactory.getTelemetryClient());
    return ErrorHandler.getHelpLink(
        HELP_NODES.UI_SERVICE_GENERATOR,
        ERROR_TYPE.INTERNAL_SERVER_ERROR,
        t('error.validatingContent')
    );
}

/**
 * Determines if the draft enabled prompt is shown.
 *
 * @param useDraftEnabled - the use draft enabled
 * @returns the boolean
 */
export function defaultOrShowDraftQuestion(useDraftEnabled?: boolean): boolean {
    return useDraftEnabled ?? true;
}

/**
 * Determines if the app gen launch prompt is shown.
 *
 * @param useLaunchGen - the use launch gen
 * @returns the boolean
 */
export function defaultOrShowAppGenLaunchQuestion(useLaunchGen?: boolean): boolean {
    return useLaunchGen ?? true;
}
