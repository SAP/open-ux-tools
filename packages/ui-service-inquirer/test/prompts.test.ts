import { getSystemSelectionPrompts, getConfigPrompts } from '../src';
import nock from 'nock';
import type { Answers, Question as YoQuestion } from 'inquirer';
import * as promptHelper from '../src/prompts/prompt-helper';
import { genContent } from './fixtures/constants';
import { PromptState } from '../src/prompts/prompt-state';
import { ObjectType } from '../src/types';
import type { YUIQuestion } from '@sap-ux/inquirer-common';
import type { AbapServiceProvider } from '@sap-ux/axios-extension';

jest.mock('../src/logger-helper');

const mockIsAppStudio = jest.fn();
jest.mock('@sap-ux/btp-utils', () => {
    return {
        ...(jest.requireActual('@sap-ux/btp-utils') as object),
        isAppStudio: () => mockIsAppStudio()
    };
});

jest.mock('../src/logger-helper', () => ({
    logger: {
        warn: jest.fn(),
        error: jest.fn()
    }
}));

jest.mock('@sap-ux/inquirer-common', () => ({
    ...(jest.requireActual('@sap-ux/inquirer-common') as {}),
    setTelemetryClient: () => jest.fn()
}));

jest.mock('@sap-ux/telemetry', () => ({
    ...(jest.requireActual('@sap-ux/telemetry') as {}),
    ClientFactory: {
        getTelemetryClient: jest.fn().mockResolvedValue({})
    }
}));
interface Question extends YoQuestion {
    when?: (answers: Answers) => boolean | Promise<boolean>;
    message?: (answers?: Answers) => string;
    choices?: () => string[];
    source?: (answers: Answers, input: string) => string[];
    additionalInfo?: () => string;
}

jest.mock('@sap-ux/system-access', () => {
    return {
        ...(jest.requireActual('@sap-ux/system-access') as object),
        createAbapServiceProvider: jest.fn().mockResolvedValue({ get: jest.fn() } as any)
    };
});

jest.mock('@sap-ux/odata-service-inquirer', () => ({
    ...(jest.requireActual('@sap-ux/odata-service-inquirer') as object),
    getSystemSelectionQuestions: jest.fn().mockResolvedValue({
        prompts: [
            {
                type: 'list',
                name: 'systemSelection',
                message: 'Select a system',
                choices: [
                    { name: 'system1', value: 'system1' },
                    { name: 'system2', value: 'system2' },
                    { name: 'system3', value: 'system3' }
                ]
            }
        ],
        answers: {
            connectedSystem: {
                backendSystem: {
                    name: 'system1'
                },
                destination: {
                    name: 'system1'
                },
                serviceProvider: {
                    getUiServiceGenerator: {
                        generate: jest.fn()
                    }
                } as unknown as AbapServiceProvider
            }
        }
    })
}));

const answersMock = {
    systemSelection: 'system1',
    transport: 'transport1',
    serviceConfig: 'serviceConfig1',
    packageInputChoice: 'manualInput'
};

describe('getSystemQuestions', () => {
    beforeAll(() => {
        nock.disableNetConnect();
    });

    afterAll(() => {
        nock.cleanAll();
        nock.enableNetConnect();
    });

    let questions;
    const q: { [key: string]: Question } = {};

    test('getSystemQuestions', async () => {
        const businessObjectMockResp = {
            name: 'I_BANKTP',
            value: {
                name: 'I_BANKTP',
                description: 'Banking',
                uri: '/sap/bc/adt/bo/behaviordefinitions/I_BANKTP'
            }
        };

        const abapCDSViewsMockResp = {
            name: 'C_GRANTORCLAIMITEMDEX',
            value: {
                name: 'C_GRANTORCLAIMITEMDEX',
                description: 'Grant Claim Item',
                uri: '/sap/bc/adt/ddic/ddl/sources/C_GRANTORCLAIMITEMDEX'
            }
        };

        jest.spyOn(promptHelper, 'getBusinessObjects').mockResolvedValue([businessObjectMockResp]);
        jest.spyOn(promptHelper, 'getAbapCDSViews').mockResolvedValue([abapCDSViewsMockResp]);
        questions = (await getSystemSelectionPrompts()).prompts;
        questions.forEach((question: any) => {
            q[question.name] = question as Question;
        });
        expect(questions).toMatchSnapshot();

        expect(q.objectType.when!({ systemSelection: 'system1' })).toEqual(true);
        expect(q.objectType.choices!()).toEqual([
            { name: 'prompts.businessObjectInterfaceLabel', value: ObjectType.BUSINESS_OBJECT },
            { name: 'prompts.abapCdsServiceLabel', value: ObjectType.CDS_VIEW }
        ]);

        const businessObjectMock = {
            name: 'I_BANKTP',
            description: 'Banking',
            uri: '/sap/bc/adt/bo/behaviordefinitions/I_BANKTP'
        };

        const abapCDSViewMock = {
            name: 'C_GRANTORCLAIMITEMDEX',
            description: 'Grant Claim Item',
            uri: '/sap/bc/adt/ddic/ddl/sources/C_GRANTORCLAIMITEMDEX'
        };
        expect(await q.businessObjectInterface.when!({ objectType: ObjectType.BUSINESS_OBJECT })).toEqual(true);
        expect(await q.businessObjectInterface.when!({ objectType: ObjectType.CDS_VIEW })).toEqual(false);
        expect(await q.businessObjectInterface.choices!()).toEqual([businessObjectMockResp]);
        expect(await q.businessObjectInterface.validate!(businessObjectMock, answersMock)).toEqual(
            'error.noGeneratorFoundBo'
        );

        expect(await q.abapCDSView.when!({ objectType: ObjectType.CDS_VIEW })).toEqual(true);
        expect(await q.abapCDSView.choices!()).toEqual([abapCDSViewsMockResp]);
        expect(await q.abapCDSView.validate!(abapCDSViewMock, answersMock)).toEqual('error.noGeneratorFoundCdsService');

        mockIsAppStudio.mockReturnValue(false);
        questions = (await getSystemSelectionPrompts()).prompts;
        questions.forEach((question: any) => {
            q[question.name] = question as Question;
        });
    });

    test('getServiceConfigQuestions', async () => {
        const getContentMock = jest.fn().mockImplementation((pckg) => {
            const serviceName = pckg === '' ? '' : 'serviceName';
            return Promise.resolve(
                JSON.stringify({
                    businessService: {
                        serviceBinding: {
                            serviceBindingName: serviceName
                        }
                    },
                    businessObject: {
                        projectionBehavior: {
                            withDraft: undefined
                        }
                    }
                })
            );
        });
        const genMock = {
            getContent: getContentMock,
            validateContent: jest.fn().mockResolvedValue({
                severity: 'OK'
            })
        } as any;
        const systemSelectionAnswers = {
            connectedSystem: {
                backendSystem: {
                    name: 'system1'
                },
                destination: {
                    Name: 'system1'
                },
                serviceProvider: {
                    get: jest.fn(),
                    getUiServiceGenerator: jest.fn().mockResolvedValue({ generate: jest.fn().mockResolvedValue({}) })
                }
            },
            objectGenerator: genMock
        } as any;

        questions = getConfigPrompts(systemSelectionAnswers, {
            useDraftEnabled: undefined,
            useLaunchGen: undefined
        }).prompts;
        questions.forEach((question) => {
            q[question.name] = question as Question;
        });
        expect(questions).toMatchSnapshot();
        expect(q.transportInputChoice.choices!()).toMatchSnapshot();
        expect(await q.serviceName.when!({ packageManual: '', packageAutocomplete: 'package' })).toEqual(true);
        expect(q.serviceName.choices!()).toEqual([{ name: 'serviceName', value: 'serviceName' }]);
        expect(q.serviceName.default()).toEqual(0);
        expect(await q.serviceName.validate!('testPackage')).toEqual(true);
        expect(q.draftEnabled.when!({})).toEqual(true);
        expect(await q.draftEnabled.validate!(true)).toEqual(true);
        expect(await q.draftEnabled.validate!(false)).toEqual(true);
        expect(q.launchAppGen.when!({})).toEqual(true);
        expect((q.launchAppGen as YUIQuestion).additionalMessages!(false)).toBeUndefined();
        expect((q.launchAppGen as YUIQuestion).additionalMessages!(true)).toEqual({
            message: 'info.appGenLaunch',
            severity: 2
        });

        const genMockValidateContent1 = {
            getContent: getContentMock,
            validateContent: jest.fn().mockResolvedValue({
                severity: 'ERROR'
            })
        } as any;
        const systemSelectionAnswers1 = {
            connectedSystem: {
                backendSystem: {
                    name: 'system1'
                },
                destination: {
                    Name: 'system1'
                },
                serviceProvider: {
                    get: jest.fn(),
                    getUiServiceGenerator: jest.fn().mockResolvedValue({ generate: jest.fn().mockResolvedValue({}) })
                }
            },
            objectGenerator: genMockValidateContent1
        } as any;

        PromptState.resetConnectedSystem();
        questions = getConfigPrompts(systemSelectionAnswers1, {
            useDraftEnabled: false,
            useLaunchGen: false
        }).prompts;
        questions.forEach((question) => {
            q[question.name] = question as Question;
        });
        expect(await q.serviceName.when!({ packageManual: 'package' })).toEqual(true);
        expect(await q.serviceName.choices!()).toEqual([{ name: 'serviceName', value: 'serviceName' }]);
        expect(await q.serviceName.validate!('testPackage')).toMatchSnapshot();
        expect(q.draftEnabled.when!({})).toEqual(false);
        expect(await q.draftEnabled.validate!(false)).toEqual('error.validatingContent');
        PromptState.resetServiceConfig();
        expect(await q.serviceName.when!({ packageManual: '', packageAutocomplete: '' })).toEqual(false);
        PromptState.resetServiceConfig();
        expect(await q.serviceName.when!({ packageManual: undefined, packageAutocomplete: undefined })).toEqual(false);
        expect(q.launchAppGen.when!({})).toEqual(false);
        const genMockValidateContent2 = {
            getContent: getContentMock,
            validateContent: jest.fn().mockImplementation(() => {
                throw new Error('error');
            })
        } as any;
        const systemSelectionAnswers2 = {
            connectedSystem: {
                backendSystem: {
                    name: 'system1'
                },
                destination: {
                    Name: 'system1'
                },
                serviceProvider: {
                    get: jest.fn(),
                    getUiServiceGenerator: jest.fn().mockResolvedValue({ generate: jest.fn().mockResolvedValue({}) })
                }
            },
            objectGenerator: genMockValidateContent2
        } as any;
        PromptState.resetConnectedSystem();
        questions = getConfigPrompts(systemSelectionAnswers2).prompts;
        questions.forEach((question) => {
            q[question.name] = question as Question;
        });
        expect(await q.serviceName.when!({ packageAutocomplete: 'package' })).toEqual(true);
        expect(await q.serviceName.validate!('testPackage')).toMatchSnapshot();
    });
});
