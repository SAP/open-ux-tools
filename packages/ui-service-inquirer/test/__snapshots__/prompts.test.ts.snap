// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSystemQuestions getServiceConfigQuestions 1`] = `
[
  {
    "choices": [Function],
    "default": [Function],
    "guiOptions": {
      "applyDefaultWhenDirty": true,
    },
    "message": "prompts.config.package.packageInputChoice.message",
    "name": "packageInputChoice",
    "type": "list",
    "validate": [Function],
    "when": [Function],
  },
  {
    "name": "packageCliExecution",
    "type": "input",
    "when": [Function],
  },
  {
    "default": [Function],
    "guiOptions": {
      "breadcrumb": true,
      "hint": "prompts.config.package.packageManual.hint",
      "mandatory": true,
    },
    "message": "prompts.config.package.packageManual.message",
    "name": "packageManual",
    "type": "input",
    "validate": [Function],
    "when": [Function],
  },
  {
    "additionalInfo": [Function],
    "guiOptions": {
      "breadcrumb": true,
      "hint": "prompts.config.package.packageAutocomplete.hint",
      "mandatory": true,
    },
    "message": "prompts.config.package.packageAutocomplete.message",
    "name": "packageAutocomplete",
    "source": [Function],
    "type": "autocomplete",
    "validate": [Function],
    "when": [Function],
  },
  {
    "choices": [Function],
    "default": [Function],
    "guiOptions": {
      "applyDefaultWhenDirty": true,
    },
    "message": "prompts.config.transport.transportInputChoice.message",
    "name": "transportInputChoice",
    "type": "list",
    "validate": [Function],
    "when": [Function],
  },
  {
    "name": "transportCliExecution",
    "type": "input",
    "when": [Function],
  },
  {
    "default": [Function],
    "message": "prompts.config.transport.transportCreated.message",
    "name": "transportCreated",
    "type": "input",
    "when": [Function],
  },
  {
    "choices": [Function],
    "default": [Function],
    "guiOptions": {
      "breadcrumb": "prompts.config.transport.common.transportRequest",
      "hint": "prompts.config.transport.common.provideTransportRequest",
    },
    "message": "prompts.config.transport.common.transportRequest",
    "name": "transportFromList",
    "type": "list",
    "when": [Function],
  },
  {
    "default": [Function],
    "filter": [Function],
    "guiOptions": {
      "breadcrumb": "prompts.config.transport.common.transportRequest",
      "hint": "prompts.config.transport.common.provideTransportRequest",
    },
    "message": [Function],
    "name": "transportManual",
    "type": "input",
    "validate": [Function],
    "when": [Function],
  },
  {
    "choices": [Function],
    "default": [Function],
    "guiOptions": {
      "breadcrumb": "PROMPT_SERVICE_NAME_BREADCRUMB",
    },
    "message": "PROMPT_SERVICE_NAME",
    "name": "serviceName",
    "type": "expand",
    "validate": [Function],
    "when": [Function],
  },
  {
    "default": true,
    "guiOptions": {
      "breadcrumb": true,
      "mandatory": true,
    },
    "message": "PROMPT_DRAFT_ENABLED",
    "name": "draftEnabled",
    "type": "confirm",
    "validate": [Function],
  },
  {
    "additionalMessages": [Function],
    "default": false,
    "guiOptions": {
      "breadcrumb": "PROMPT_LAUNCH_APP_GEN_BREADCRUMB",
    },
    "message": "PROMPT_LAUNCH_APP_GEN",
    "name": "launchAppGen",
    "type": "confirm",
  },
]
`;

exports[`getSystemQuestions getServiceConfigQuestions 2`] = `
[
  {
    "name": "choices.transport.enterManually",
    "value": "EnterManualChoice",
  },
  {
    "name": "choices.common.listExisting",
    "value": "ListExistingChoice",
  },
  {
    "name": "choices.transport.createNew",
    "value": "CreateNewChoice",
  },
]
`;

exports[`getSystemQuestions getServiceConfigQuestions 3`] = `
ValidationLink {
  "link": {
    "icon": "data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTIiIGhlaWdodD0iMTIiIHZpZXdCb3g9IjAgMCAxMiAxMiIgZmlsbD0ibm9uZSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KPGcgY2xpcC1wYXRoPSJ1cmwoI2NsaXAwXzM0NTRfMTc4MDUpIj4KPHBhdGggZmlsbC1ydWxlPSJldmVub2RkIiBjbGlwLXJ1bGU9ImV2ZW5vZGQiIGQ9Ik0xMSA3LjI0MjIzVjkuNUMxMSA5LjY4OTM5IDEwLjg5MyA5Ljg2MjUyIDEwLjcyMzYgOS45NDcyMUw3LjcyMzYxIDExLjQ0NzJDNy41OTkyNSAxMS41MDk0IDcuNDU0NjIgMTEuNTE3IDcuMzI0NDQgMTEuNDY4MkwzLjQ5MDE4IDEwLjAzMDNMMC42NTgxMTQgMTAuOTc0M0MwLjUwNTY0IDExLjAyNTIgMC4zMzgwMjkgMTAuOTk5NiAwLjIwNzY0NSAxMC45MDU2QzAuMDc3MjYwNiAxMC44MTE2IDAgMTAuNjYwNyAwIDEwLjVWMi41QzAgMi4yODQ3OCAwLjEzNzcxNSAyLjA5MzcyIDAuMzQxODg2IDIuMDI1NjZMMy4zNDE4OSAxLjAyNTY2QzMuNDQ3OTcgMC45OTAyOTYgMy41NjI4NiAwLjk5MTUxNyAzLjY2ODE3IDEuMDI5MTNMNC41NDkyNSAxLjM0MzhDNC4zOTAxOSAxLjYzNDYzIDQuMjYyMyAxLjk0NDk0IDQuMTcwMDkgMi4yNzAyNUw0IDIuMjA5NVY5LjE1MzVMNyAxMC4yNzg1VjcuNzQzOTRDNy4zMTg0MSA3Ljg1NjQ4IDcuNjUzMjcgNy45MzQyMSA4IDcuOTcyNTRWMTAuMTkxTDEwIDkuMTkwOThWNy43NDM5NEMxMC4zNTYgNy42MTgxIDEwLjY5MTUgNy40NDg3MyAxMSA3LjI0MjIzWk0xIDkuODA2MjlWMi44NjAzOEwzIDIuMTkzNzFWOS4xMzk2MkwxIDkuODA2MjlaIiBmaWxsPSIjQzVDNUM1Ii8+CjxwYXRoIGZpbGwtcnVsZT0iZXZlbm9kZCIgY2xpcC1ydWxlPSJldmVub2RkIiBkPSJNNSAzLjQ2NjIyQzUgNC4yNTM4IDUuMjYyNjcgNC45ODAwMyA1LjcwNTE4IDUuNTYyMDlMNS4zNzc1OCA3LjQyTDcuMjg0NzEgNi43MjU4NkM3LjY1MzQ1IDYuODU5NTUgOC4wNTEzMiA2LjkzMjQ0IDguNDY2MjIgNi45MzI0NEMxMC4zODA2IDYuOTMyNDQgMTEuOTMyNCA1LjM4MDU2IDExLjkzMjQgMy40NjYyMkMxMS45MzI0IDEuNTUxODggMTAuMzgwNiAwIDguNDY2MjIgMEM2LjU1MTg4IDAgNSAxLjU1MTg4IDUgMy40NjYyMlpNOCAxLjVDOCAxLjIyMzg2IDguMjIzODYgMSA4LjUgMUM4Ljc3NjE0IDEgOSAxLjIyMzg2IDkgMS41QzkgMS43NzYxNCA4Ljc3NjE0IDIgOC41IDJDOC4yMjM4NiAyIDggMS43NzYxNCA4IDEuNVpNOC41IDNDOC4yMjM4NiAzIDggMy4yMjM4NiA4IDMuNVY1LjVDOCA1Ljc3NjE0IDguMjIzODYgNiA4LjUgNkM4Ljc3NjE0IDYgOSA1Ljc3NjE0IDkgNS41VjMuNUM5IDMuMjIzODYgOC43NzYxNCAzIDguNSAzWiIgZmlsbD0iI0M1QzVDNSIvPgo8L2c+CjxkZWZzPgo8Y2xpcFBhdGggaWQ9ImNsaXAwXzM0NTRfMTc4MDUiPgo8cmVjdCB3aWR0aD0iMTIiIGhlaWdodD0iMTIiIGZpbGw9IndoaXRlIi8+CjwvY2xpcFBhdGg+CjwvZGVmcz4KPC9zdmc+Cg==",
    "text": "guidedAnswers.validationErrorHelpText",
    "url": "https://ga.support.sap.com/dtp/viewer/index.html#/tree/3046/actions/63068",
  },
  "message": "ERROR_VALIDATING_CONTENT",
}
`;

exports[`getSystemQuestions getServiceConfigQuestions 4`] = `
ValidationLink {
  "link": {
    "icon": "data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTIiIGhlaWdodD0iMTIiIHZpZXdCb3g9IjAgMCAxMiAxMiIgZmlsbD0ibm9uZSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KPGcgY2xpcC1wYXRoPSJ1cmwoI2NsaXAwXzM0NTRfMTc4MDUpIj4KPHBhdGggZmlsbC1ydWxlPSJldmVub2RkIiBjbGlwLXJ1bGU9ImV2ZW5vZGQiIGQ9Ik0xMSA3LjI0MjIzVjkuNUMxMSA5LjY4OTM5IDEwLjg5MyA5Ljg2MjUyIDEwLjcyMzYgOS45NDcyMUw3LjcyMzYxIDExLjQ0NzJDNy41OTkyNSAxMS41MDk0IDcuNDU0NjIgMTEuNTE3IDcuMzI0NDQgMTEuNDY4MkwzLjQ5MDE4IDEwLjAzMDNMMC42NTgxMTQgMTAuOTc0M0MwLjUwNTY0IDExLjAyNTIgMC4zMzgwMjkgMTAuOTk5NiAwLjIwNzY0NSAxMC45MDU2QzAuMDc3MjYwNiAxMC44MTE2IDAgMTAuNjYwNyAwIDEwLjVWMi41QzAgMi4yODQ3OCAwLjEzNzcxNSAyLjA5MzcyIDAuMzQxODg2IDIuMDI1NjZMMy4zNDE4OSAxLjAyNTY2QzMuNDQ3OTcgMC45OTAyOTYgMy41NjI4NiAwLjk5MTUxNyAzLjY2ODE3IDEuMDI5MTNMNC41NDkyNSAxLjM0MzhDNC4zOTAxOSAxLjYzNDYzIDQuMjYyMyAxLjk0NDk0IDQuMTcwMDkgMi4yNzAyNUw0IDIuMjA5NVY5LjE1MzVMNyAxMC4yNzg1VjcuNzQzOTRDNy4zMTg0MSA3Ljg1NjQ4IDcuNjUzMjcgNy45MzQyMSA4IDcuOTcyNTRWMTAuMTkxTDEwIDkuMTkwOThWNy43NDM5NEMxMC4zNTYgNy42MTgxIDEwLjY5MTUgNy40NDg3MyAxMSA3LjI0MjIzWk0xIDkuODA2MjlWMi44NjAzOEwzIDIuMTkzNzFWOS4xMzk2MkwxIDkuODA2MjlaIiBmaWxsPSIjQzVDNUM1Ii8+CjxwYXRoIGZpbGwtcnVsZT0iZXZlbm9kZCIgY2xpcC1ydWxlPSJldmVub2RkIiBkPSJNNSAzLjQ2NjIyQzUgNC4yNTM4IDUuMjYyNjcgNC45ODAwMyA1LjcwNTE4IDUuNTYyMDlMNS4zNzc1OCA3LjQyTDcuMjg0NzEgNi43MjU4NkM3LjY1MzQ1IDYuODU5NTUgOC4wNTEzMiA2LjkzMjQ0IDguNDY2MjIgNi45MzI0NEMxMC4zODA2IDYuOTMyNDQgMTEuOTMyNCA1LjM4MDU2IDExLjkzMjQgMy40NjYyMkMxMS45MzI0IDEuNTUxODggMTAuMzgwNiAwIDguNDY2MjIgMEM2LjU1MTg4IDAgNSAxLjU1MTg4IDUgMy40NjYyMlpNOCAxLjVDOCAxLjIyMzg2IDguMjIzODYgMSA4LjUgMUM4Ljc3NjE0IDEgOSAxLjIyMzg2IDkgMS41QzkgMS43NzYxNCA4Ljc3NjE0IDIgOC41IDJDOC4yMjM4NiAyIDggMS43NzYxNCA4IDEuNVpNOC41IDNDOC4yMjM4NiAzIDggMy4yMjM4NiA4IDMuNVY1LjVDOCA1Ljc3NjE0IDguMjIzODYgNiA4LjUgNkM4Ljc3NjE0IDYgOSA1Ljc3NjE0IDkgNS41VjMuNUM5IDMuMjIzODYgOC43NzYxNCAzIDguNSAzWiIgZmlsbD0iI0M1QzVDNSIvPgo8L2c+CjxkZWZzPgo8Y2xpcFBhdGggaWQ9ImNsaXAwXzM0NTRfMTc4MDUiPgo8cmVjdCB3aWR0aD0iMTIiIGhlaWdodD0iMTIiIGZpbGw9IndoaXRlIi8+CjwvY2xpcFBhdGg+CjwvZGVmcz4KPC9zdmc+Cg==",
    "text": "guidedAnswers.validationErrorHelpText",
    "url": "https://ga.support.sap.com/dtp/viewer/index.html#/tree/3046/actions/63068",
  },
  "message": "ERROR_VALIDATING_CONTENT",
}
`;

exports[`getSystemQuestions getSystemQuestions 1`] = `
[
  {
    "choices": [
      {
        "name": "system1",
        "value": "system1",
      },
      {
        "name": "system2",
        "value": "system2",
      },
      {
        "name": "system3",
        "value": "system3",
      },
    ],
    "message": "Select a system",
    "name": "systemSelection",
    "type": "list",
  },
  {
    "choices": [Function],
    "default": "",
    "guiOptions": {
      "breadcrumb": true,
    },
    "message": "MESSAGE_OBJECT_TYPE",
    "name": "objectType",
    "type": "list",
    "when": [Function],
  },
  {
    "choices": [Function],
    "default": "",
    "guiOptions": {
      "applyDefaultWhenDirty": true,
      "breadcrumb": true,
    },
    "message": "MESSAGE_BUSINESS_OBJECT_INTERFACE",
    "name": "businessObjectInterface",
    "type": "list",
    "validate": [Function],
    "when": [Function],
  },
  {
    "choices": [Function],
    "default": "",
    "guiOptions": {
      "applyDefaultWhenDirty": true,
      "breadcrumb": true,
    },
    "message": "MESSAGE_ABAP_CDS_SERVICE",
    "name": "abapCDSView",
    "type": "list",
    "validate": [Function],
    "when": [Function],
  },
]
`;
